### ISSUE TITLE
Dataset creation requires tuple, list treated differently

### ISSUE THREAD

**Issue by mcgibbon** [pydata/xarray#929]
Take the Dataset creation example:

``` python
In [35]: ds = xr.Dataset({'temperature': (['x', 'y', 'time'],  temp),
   ....:                  'precipitation': (['x', 'y', 'time'], precip)},
   ....:                 coords={'lon': (['x', 'y'], lon),
   ....:                         'lat': (['x', 'y'], lat),
   ....:                         'time': pd.date_range('2014-09-06', periods=3),
   ....:                         'reference_time': pd.Timestamp('2014-09-05')})
   ....: 
```

if the tuple (['x', 'y', 'time'], temp) is replaced with a list [['x', 'y', 'time'], temp], the behavior changes in very strange ways. The resulting Dataset will then have a _coordinate_ variable temperature whose dimensions are ('temperature', 'x', 'y', 'time'). Printing temperature shows that the ['x', 'y', 'time'] part has been interpreted as data rather than metadata. It seems to be impossible to access the data in the resulting temperature coordinate by indexing.

This might be intentional (since one could actually want to pass in data that is stored as a list), but it may be better to do some sanity checking when a list is passed to figure out whether the list is data or as above. If no change is made, then this feature should probably be pointed out in the documentation.

**Comment by fmaussion** [comment]
> it may be better to do some sanity checking when a list is passed to figure out whether the list is data or as above

I guess that this is the difficult part. How to know if the list is data or "as above"?

**Comment by mcgibbon** [comment]
A simple way is to try creating a Data Array by assuming it's metadata, and if it fails then assume it's data. This will probably "work" too often on lists it shouldn't, however. A better heuristic is that if the list is of length 2 or 3, and the first element is an iterable of strings, and the second element contains data (is a list or numpy array), and the optional third element is a map/dictionary, then the list is probably metadata. You might also require that the second element isn't the same length as the first element, in the case that they're both lists (in case someone wants a 2x2 DataArray of string labels that they're constructing from a list of lists).

**Comment by shoyer** [comment]
This difference is certainly intentional. You're right that we currently don't explicitly state how we handle lists or numpy arrays in the docs -- that should be fixed.

At one point, we did have heuristics that did this sort of fallback logic (allowing lists, and checking the length of the tuple/list). But we eventually removed these checks, because this sort of value dependent behavior makes code difficult to predict. At least now it always behaves consistently when given a list or a tuple.
